### PR TITLE
Add experimental support for Stackdriver logging to log package

### PR DESCRIPTION
**Pull request by douglas-reid**
This PR introduces new functionality into the `log` package to support better integration with Google Cloud Logging (aka Stackdriver). The existing log support has a few drawbacks, including:

-  poor alignment with Stackdriver conventions (level v. severity, severity level names, time v. timestamp, etc.)
- no way to log directly to Stackdriver

This PR introduces two bits:

- A new logs format that is Stackdriver-compatible
- A "TeeToStackdriver" option that will allow direct writes to the Cloud Logging API

Example CLI logs for Pilot with Stackdriver-compatible format:

```json
{"severity":"Info","timestamp":"2021-02-05T23:41:34.688920101Z","logger":"ads","message":"Push debounce stable[2] 2: 100.179249ms since last change, 177.057621ms since last push, full=false"}
{"severity":"Info","timestamp":"2021-02-05T23:41:34.689148455Z","logger":"ads","message":"XDS: Incremental Pushing:2021-02-05T23:41:32Z/1 ConnectedEndpoints:0 Version:2021-02-05T23:41:32Z/1"}
{"severity":"Info","timestamp":"2021-02-05T23:41:35.268693809Z","logger":"ads","message":"Push debounce stable[3] 1: 100.139099ms since last change, 100.138885ms since last push, full=false"}
{"severity":"Info","timestamp":"2021-02-05T23:41:35.268959092Z","logger":"ads","message":"XDS: Incremental Pushing:2021-02-05T23:41:32Z/1 ConnectedEndpoints:0 Version:2021-02-05T23:41:32Z/1"}
{"severity":"Info","timestamp":"2021-02-05T23:41:40.01736317Z","logger":"ads","message":"Push debounce stable[4] 2: 100.231342ms since last change, 194.19688ms since last push, full=false"}
{"severity":"Info","timestamp":"2021-02-05T23:41:40.017491712Z","logger":"ads","message":"XDS: Incremental Pushing:2021-02-05T23:41:32Z/1 ConnectedEndpoints:0 Version:2021-02-05T23:41:32Z/1"}
{"severity":"Info","timestamp":"2021-02-05T23:41:41.636460945Z","logger":"klog","message":"successfully acquired lease istio-system/istio-leader"}
{"severity":"Info","timestamp":"2021-02-05T23:41:41.641106133Z","message":"Starting ingress controller"}
```

Example Stackdriver view of logs:

image

### CODE REVIEW

**Comment by douglas-reid** [comment]
Gentle ping for review.

File: log/config.go [hunk 1]
```diff
@@ -105,27 +105,45 @@ func init() {
...
-		EncodeLevel:    zapcore.LowercaseLevelEncoder,
-		EncodeCaller:   zapcore.ShortCallerEncoder,
-		EncodeDuration: zapcore.StringDurationEncoder,
-		EncodeTime:     formatDate,
-	}
-
 	var enc zapcore.Encoder
-	if options.JSONEncoding {
+	if options.useStackdriverFormat {
+		encCfg := zapcore.EncoderConfig{
```
Comment (howardjohn) [inline log/config.go:110]: can you link to the format? I assume https://cloud.google.com/logging/docs/reference/v2/rest/v2/LogEntry?

File: log/config.go [hunk 2]
```diff
@@ -105,27 +105,45 @@ func init() {
...
+			TimeKey:        "timestamp",
+			LevelKey:       "severity",
+			NameKey:        "logger",
```
Comment (howardjohn) [inline log/config.go:113]: I don't see this key defined in the format above?

Comment (douglas-reid) [reply log/config.go:113]: Logger isn't in the LogEntry proto; that is correct.

However, it ensures that the Scope name is consistently placed in the jsonPayload.logger field. This matches the code for the direct-to-stackdriver bits (here: https://github.com/istio/pkg/pull/318/files#diff-468565f13023bc1ab3a3eb1dd6930ae69140cb051c3fedc7f2b9f9335319ae23R133). 

So, that bit is more about consistency. If you think we should use a different field name, let's do that.

File: log/config.go (hunk 1 above)
Comment (douglas-reid) [reply log/config.go:111]: done.

**Comment by douglas-reid** [comment]
Any further items that should be addressed ?

**Review by ostromart — APPROVED** [review]

**Review by howardjohn — APPROVED** [review]

**Pull request merged**